Permet de signaler un message comme spam potentiel

Situphen opened the pull request:
Permet de signaler un message comme spam potentiel et créer automatiquement une alerte lorsque que ce message est modifié.

Cadeau pour @arobase-che de la part de @AmauryCarrade et moi même !

### Contrôle qualité

Il y a pas mal d'instructions de QA, mais elles sont assez rapides à exécuter.

Avant tout, appliquez les migrations puis lancez le site et actualisez de force une page quelconque pour vider le cache du navigateur.

```console
make migrate-db && make zmd-start && make run
```

Ensuite, ouvrez deux onglets indépendants de ZdS (par exemple, un en session normale et un en session de navigation privée). Connectez-vous avec un utilisateur normal sur l'un, et avec un utilisateur ayant des droits de _staff_ sur l'autre.

_Dans tout ce qui suit, et sauf mention contraire, les modifications doivent être réelles, dans le sens que le texte doit vraiment changer. Il ne suffit pas de soumettre le formulaire de modification sans changer le contenu._

1. Trouvez sur le forum un message envoyé par votre utilisateur normal. Modifiez ce message avec le compte normal. Depuis le compte _staff_, vérifiez qu'aucune alerte n'est levée.
2. Depuis le compte staff, marquez le message comme spam potentiel en cliquant sur l'icône en forme de panneau d'avertissement en bas à droite du message. Le message doit virer au rouge, avec une en-tête indiquant qu'il est soupçonné de spam.
3. Depuis le compte staff, modifiez le message. Vérifiez qu'aucune alerte n'est levée.
4. Depuis le compte utilisateur normal, modifiez le message. Vérifiez qu'une alerte est levée.
5. Depuis le compte utilisateur normal, et sans avoir résolu l'alerte, modifiez le message. Vérifiez qu'aucune _nouvelle_ alerte n'est levée, mais que l'ancienne alerte est actualisée à la date de la dernière modification.
6. Depuis le compte staff, résolvez l'alerte. Vérifiez au passage qu'il n'y a pas de zone de texte pour envoyer un message, contrairement aux alertes classiques, vu que l'alerte est envoyée par un robot.
6. Depuis le compte utilisateur normal, modifiez le message _sans modifier le texte_. Vérifiez qu'aucune alerte n'est levée.
7. Depuis le compte utilisateur normal, modifiez le message. Vérifiez qu'une _nouvelle_ alerte est levée.
8. Depuis le compte utilisateur normal, signalez un autre message. Vérifiez que si vous résolvez l'alerte ainsi créée, le champ pour envoyer un message à l'auteur de l'alerte est bien présent (car l'alerte est envoyée par une personne physique).

Situphen committed:
Permet de signaler un message comme spam potentiel

Situphen committed:
Oups

AmauryCarrade committed:
Amélioration du style graphique et de l'UX des spams potentiels.

- Fond rouge et non vert.
- Icône adaptée (alerte, et non tick).
- Icône plus discrète quand le mode n'est pas activé.
- Textes plus explicites pour comprendre ce que ça fait.

AmauryCarrade committed:
Alerts does not stack if a potential spam is updated multiple times.

AmauryCarrade committed:
Le marquage d'un message comme spam n'est visible que pour le staff.

Aussi, ne remonte pas d'alerte si l'auteur n'est pas l'auteur original.

AmauryCarrade committed:
Tests.

Aussi, reformulation de messages et style de code.

AmauryCarrade committed:
Retire le formulaire d'envoi de message à l'auteur de l'alerte, si automatique.

AmauryCarrade committed:
Meilleur message pour les alertes automatiques sans message envoyé.

AmauryCarrade committed:
Migration oubliée

AmauryCarrade commented:
À Travis près qui râle un peu, tout est bon ici. Les instructions de QA ont été ajoutées au message initial.

AmauryCarrade committed:
Corrige un test et un style de code

- Corrige un test qui s'appuyait sur du code HTML qui n'était plus unique avec cette
  PR pour fonctionner (pour détecter le nombre de messages alertables). Le test a été
  mis à jour avec une méthode fonctionnelle et plus pérenne. Aussi, un commentaire
  devenu faux suite à une mise à jour a été corrigé.
- Corrige une ligne trop longue.

AmauryCarrade committed:
Correction du test

- La méthode utilisée pour marquer une alerte comme traitée n'était pas la bonne.
  Utiliser la méthode normale est toujours mieux…

AmauryCarrade committed:
Merge branch 'dev' into potential-spam

philippemilink commented:
Rapport de QA
--------------------

**OK** :heavy_check_mark: 

Mais j'ai une remarque: dans la barre qui s'affiche sous le message potentiellement spammeur, il est écrit: _il y a x minutes par **admin**_. Pourquoi admin, alors que c'est un message automatique ? Il n'y a pas un compte bot pour ce genre d'action ?

Un rebase et c'est bon.

Situphen committed:
Merge branch 'dev' into potential-spam

Situphen commented:
> Pourquoi admin, alors que c'est un message automatique ? Il n'y a pas un compte bot pour ce genre d'action ?

Sur la version locale, le compte bot est admin. Sur la production, ce sera bien entendu Clem :)
